File: Moose’s Algorithm/code.java
import java.util.Arrays;

public class MooseAlgorithm {
    public static int[] mooseAlgorithm(int[] arr) {
        int n = arr.length;
        int[] dp = new int[n];
        Arrays.fill(dp, Integer.MAX_VALUE);
        dp[0] = arr[0];
        
        for (int i = 1; i < n; i++) {
            for (int j = i - 1; j >= 0; j--) {
                dp[i] = Math.min(dp[i], Math.min(arr[i], dp[j]));
            }
        }
        
        return dp;
    }
    
    public static void main(String[] args) {
        int[] arr = {4, 2, 7, 1, 3, 8, 5};
        int[] result = mooseAlgorithm(arr);
        
        for (int i : result) {
            System.out.print(i + " ");
        }
        System.out.println();
    }
}

File: Moose’s Algorithm/code.py
def moose_algorithm(arr):
    n = len(arr)
    dp = [float('inf')] * n
    dp[0] = arr[0]
    
    for i in range(1, n):
        for j in range(i - 1, -1, -1):
            dp[i] = min(dp[i], min(arr[i], dp[j]))
    
    return dp

arr = [4, 2, 7, 1, 3, 8, 5]
result = moose_algorithm(arr)
print(" ".join(map(str, result)))
